ibis/expr/types/relations.py

```diff
@@ -1828,33 +1857,52 @@ def relabel(
         │ PAL0708   │             1 │ Adelie Penguin (Pygoscelis adeliae) │
         └───────────┴───────────────┴─────────────────────────────────────┘
         """
-        observed = set()
+        if isinstance(method, Mapping):
+            substitutions.update(method)
+            method = None
+
+        # A mapping from old_name -> renamed expr
+        renamed = {}
+
+        if substitutions:
+            schema = self.schema()
+            for new_name, old_name in substitutions.items():
+                col = self[old_name]
+                if old_name not in renamed:
+                    renamed[old_name] = col.name(new_name)
+                else:
+                    raise ValueError(
+                        "duplicate new names passed for renaming {old_name!r}"
+                    )
 
-        if isinstance(substitutions, Mapping):
-            rename = substitutions.get
-        elif substitutions in {"snake_case", "ALL_CAPS"}:
+        if method is None:
+
+            def rename(c):
+                return None
+
+        elif isinstance(method, str) and method in {"snake_case", "ALL_CAPS"}:
```

Comment:
consider extracting this out into a separate function, for easier unit testing (this logic is complex) and for a slimmer `def rename()`

Reply:
The logic is fully tested in our existing test suite. Keeping it inline in the method that uses it [avoids indirection](https://matthewrocklin.com/blog/work/2019/06/23/avoid-indirection) making it easier for a reader to see what's going on.

Reply:
Ah, I see that now. Thanks! I added a comment inline there, I think this could do with apostrophes, other whitespace chars besides spaces, and numbers.